storage/src/commit_log/mod.rs

```diff
@@ -0,0 +1,430 @@
+    MissingCommitLog { name: &'static str },
+    #[fail(display = "Failed to read record at {}", location)]
+    ReadError { location: Location },
+    #[fail(display = "Failed to read record data corrupted")]
+    CorruptData,
+}
+
+impl From<SchemaError> for CommitLogError {
+    fn from(error: SchemaError) -> Self {
+        CommitLogError::SchemaError { error }
+    }
+}
+
+impl From<io::Error> for CommitLogError {
+    fn from(error: io::Error) -> Self {
+        CommitLogError::IOError { error }
+    }
+}
+
+impl slog::Value for CommitLogError {
+    fn serialize(
+        &self,
+        _record: &slog::Record,
+        key: slog::Key,
+        serializer: &mut dyn slog::Serializer,
+    ) -> slog::Result {
+        serializer.emit_arguments(key, &format_args!("{}", self))
+    }
+}
+
+type ByteLimit = usize;
+type ItemCount = u32;
+
+/// Precisely identifies location of a record in a commit log.
+#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
+pub struct Location(pub u64, pub ByteLimit);
+
+impl Location {
+    #[inline]
+    pub fn is_consecutive(&self, prev: &Location) -> bool {
+        (prev.0 < self.0) && (self.0 - prev.0 == 1)
+    }
+}
+
+impl fmt::Display for Location {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_fmt(format_args!("Location({},{})", self.0, self.1))
+    }
+}
+
+impl BincodeEncoded for Location {}
+
+/// Range of values to get from a commit log
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct Range(pub u64, pub ByteLimit, pub ItemCount);
+
+/// Implement this trait for a commit log engine.
+pub trait CommitLogWithSchema<S: CommitLogSchema> {
+    /// Append new record to a commit log.
+    fn append(&self, value: &S::Value) -> Result<Location, CommitLogError>;
+
+    /// Retrieve a stored record.
+    fn get(&self, location: &Location) -> Result<S::Value, CommitLogError>;
+}
+
+impl<S: CommitLogSchema> CommitLogWithSchema<S> for CommitLogs {
+    fn append(&self, value: &S::Value) -> Result<Location, CommitLogError> {
+        let cl = self
+            .cl_handle(S::name())
+            .ok_or(CommitLogError::MissingCommitLog { name: S::name() })?;
+        let mut cl = cl.write().expect("Write lock failed");
+        let bytes = value.encode()?;
+        let out = cl.append_msg(&bytes)?;
+
+        Ok(Location(out.0, out.1))
+    }
+
+    fn get(&self, location: &Location) -> Result<S::Value, CommitLogError> {
+        let cl = self
+            .cl_handle(S::name())
+            .ok_or(CommitLogError::MissingCommitLog { name: S::name() })?;
+        let cl = cl.read().expect("Read lock failed");
+        let bytes = cl.read(location.0, location.1)?;
+        let value = S::Value::decode(&bytes)?;
+        Ok(value)
+    }
+}
+
+pub fn fold_consecutive_locations(locations: &[Location]) -> Vec<Range> {
+    if locations.is_empty() {
+        Vec::with_capacity(0)
+    } else {
+        let mut ranges = vec![];
```

Comment:
@mambisi 
initialize `vec![]` please with `Vec::with_capacity(locations.len)` then the `ranges.push(range);` does not need to reallocate all the time on every push